src/backends/onnx/onnx_backend.h

```diff
@@ -0,0 +1,118 @@
+//    notice, this list of conditions and the following disclaimer in the
+//    documentation and/or other materials provided with the distribution.
+//  * Neither the name of NVIDIA CORPORATION nor the names of its
+//    contributors may be used to endorse or promote products derived
+//    from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+// OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#pragma once
+
+#include <NvInfer.h>
+#include <core/session/onnxruntime_c_api.h>
+#include "src/core/backend.h"
+#include "src/core/model_config.pb.h"
+#include "src/core/scheduler.h"
+#include "src/core/status.h"
+
+namespace nvidia { namespace inferenceserver {
+
+class OnnxBackend : public InferenceBackend {
+ public:
+  OnnxBackend() = default;
+  OnnxBackend(OnnxBackend&&) = default;
+
+  Status Init(const std::string& path, const ModelConfig& config);
+
+  // Create a context for execution for each instance for the
+  // serialized plans specified in 'models'.
+  Status CreateExecutionContexts(
+      OrtEnv* env, const std::unordered_map<std::string, std::string>& paths);
+  Status CreateExecutionContext(
+      const std::string& instance_name, const int gpu_device, OrtEnv* env,
+      OrtSessionOptions* base_session_options,
+      const std::unordered_map<std::string, std::string>& paths);
+
+ private:
+  // Run model on the context associated with 'runner_idx' to
+  // execute for one or more requests.
+  void Run(
+      uint32_t runner_idx, std::vector<Scheduler::Payload>* payloads,
+      std::function<void(Status)> OnCompleteQueuedPayloads);
+
+ private:
+  DISALLOW_COPY_AND_ASSIGN(OnnxBackend);
+  friend std::ostream& operator<<(std::ostream&, const OnnxBackend&);
+
+  // For each model instance there is a context.
+  struct Context {
+    // GPU device number that indicates that no gpu is available for a
+    // context (which is an invalid state since TensorRT requires a
+    // GPU).
+    static constexpr int NO_GPU_DEVICE = -1;
+
+    // Max batch size value that indicates batching is not supported.
+    static constexpr int NO_BATCHING = 0;
+
+    Context(
+        const std::string& name, const int gpu_device,
+        const int max_batch_size);
+    Context(Context&& o);
```

Comment:
Get rid of this move constructor (and add DISALLOW_MOVE) and instead use std::unique_ptr<Context> for contexts_ vector. See plan_bundle.h (all backends should make this change but some have not yet).